File: fe/src/api/file.api.ts
import adminApiClient from '@/lib/admin-api-client'
import type { ApiResponse, FileUploadResponse, FileListResponse } from '@/types/api'

const FILE_BASE_URL = '/api/files'

export const fileApi = {
  // Upload file to MinIO with presigned URL
  upload: async (file: File): Promise<FileUploadResponse> => {
    const formData = new FormData()
    formData.append('file', file)
    
    const response = await adminApiClient.post<ApiResponse<FileUploadResponse>>(
      `${FILE_BASE_URL}/upload/presigned`,
      formData,
      {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      }
    )
    return response.data.data
  },

  // List all files with presigned URLs
  list: async (): Promise<FileListResponse> => {
    const response = await adminApiClient.get<ApiResponse<FileListResponse>>(`${FILE_BASE_URL}/list`)
    return response.data.data
  },

  // Delete file
  delete: async (filename: string): Promise<void> => {
    await adminApiClient.delete<ApiResponse<void>>(`${FILE_BASE_URL}/delete/${filename}`)
  },
}

File: be/core/src/main/java/com/example/spring_ecom/controller/api/file/FileAPI.java
package com.example.spring_ecom.controller.api.file;

import com.example.spring_ecom.controller.api.file.model.FileListResponse;
import com.example.spring_ecom.controller.api.file.model.FileUploadResponse;
import com.example.spring_ecom.core.response.ApiResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.core.io.Resource;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

@Tag(name = "File", description = "File management API using MinIO")
@RequestMapping("/v1/api/files")
public interface FileAPI {

    @Operation(summary = "Upload file to MinIO (returns public URL)")
    @PostMapping(value = "/upload", consumes = "multipart/form-data")
    ApiResponse<FileUploadResponse> uploadFile(
            @Parameter(description = "File to upload") @RequestParam("file") MultipartFile file
    );

    @Operation(summary = "Upload file to MinIO (returns presigned URL with expiry)")
    @PostMapping(value = "/upload/presigned", consumes = "multipart/form-data")
    ApiResponse<FileUploadResponse> uploadFileWithPresignedUrl(
            @Parameter(description = "File to upload") @RequestParam("file") MultipartFile file,
            @Parameter(description = "Expiry time in seconds (default: 3600)") 
            @RequestParam(value = "expiry", required = false) Integer expirySeconds
    );

    @Operation(summary = "Get presigned URL for existing file")
    @GetMapping("/presigned/{filename}")
    ApiResponse<String> getPresignedUrl(
            @Parameter(description = "Filename") @PathVariable String filename,
            @Parameter(description = "Expiry time in seconds (default: 3600)") 
            @RequestParam(value = "expiry", required = false) Integer expirySeconds
    );

    @Operation(summary = "List all files in bucket")
    @GetMapping("/list")
    ApiResponse<FileListResponse> listFiles();

    @Operation(summary = "Download file by filename")
    @GetMapping("/download/{filename}")
    ResponseEntity<Resource> downloadFile(
            @Parameter(description = "Filename to download") @PathVariable String filename
    );

    @Operation(summary = "Delete file by filename")
    @DeleteMapping("/delete/{filename}")
    ApiResponse<String> deleteFile(
            @Parameter(description = "Filename to delete") @PathVariable String filename
    );
}
